Updates Interactors docs with revised copy (#798)
thefrontside/bigtest: website/docs/interactors/4-write-your-own.md

:::info Heads up

There are new and exciting changes in the works for the `createInteractor()` API. Those changes will simplify the syntax you see in the examples below, but the refactoring process will be easy to adapt to. We will share clear instructions in your console on how you can update to the new syntax so you can use Interactors today without worrying about tomorrow’s easy-to-perform updates.

:::

## Writing your first interactor

In this tutorial, we will create our own `TextField` Interactor. Although BigTest has this type of interactor built-in, recreating it is a great way to learn all the pieces that make up an interactor while using familiar examples.

There are four things to decide when creating an interactor:
1. What to name and label the interactor
2. Which HTML element or elements to target
3. The locator and filters, which helps users be able to narrow down the element they want to reference
4. The action or actions that a test should `perform` when using the interactor (like `click`)

> In this case, let's call our Interactor ‘MyTextField’, and we’ll label it as ‘my-textfield-interactor’ to distinguish it from the build-in one. Let’s say that MyTextField will target a regular input with a custom class `.my-input`, which means its HTML selector would be `input[type=text].my-input`. For locator we’ll use the input’s placeholder, and we’ll define a value filter. Finally, we’ll define a `fillIn` action for MyTextField interactor. Putting this together, we get the following definition of an Interactor:

Reviewer: > build-in

built-in

> In this case, let's call our Interactor ‘MyTextField’, and we’ll label it as ‘my-textfield-interactor’ to distinguish it from the built-in one. Let’s say that MyTextField will target a regular input with a custom class `.my-input`, which means its HTML selector would be `input[type=text].my-input`. For locator we’ll use the input’s placeholder, and we’ll define a value filter. Finally, we’ll define a `fillIn` action for MyTextField interactor. Putting this together, we get the following definition of an Interactor:

Reviewer: 🤔 i missed this. i don't know why the `custom class '.my-input'` part was added. it's not implemented in the example. nor is the `MyTextField` interactor name.

}
```

Or you could use destructuring to make it a bit shorter:

```js
actions: {
  click: ({ perform }) => perform(element => element.click());
}
```

The former syntax is necessary if you want to write an action that delegates to the actions of other interactors. For example, imagine that you want to create a form interactor that has a submit action. You could take this approach:

```js
import { Button, createInteractor } from 'bigtest';

export const Form = createInteractor<HTMLFormElement>('form')({
  selector: 'form',
  actions: {
	async submit(interactor){
  	  await interactor.find(Button('Submit')).click();
	}

>   }

Reviewer: i like the 2-space tabbing and that's what we've been using everywhere else in the docs

})
```

There’s two peculiarities about this example. First, notice we’re not using `perform` which means that we’re not performing an action on any element directly. Secondly, within the submit action definition you’ll notice that we’re using `find` to access another Interactor and calling its action. That’s what we mean with ‘delegating’ an action.

Why would you want to delegate actions within an Interactor action definition? Well, it can help you avoid repeating yourself and to keep Interactors as black boxes. For example, let’s say we wanted to test that a user submits a form, without delegation we can reach in in the Form Interactor and issue a click action:

```js

> Form.find(Button('Submit')).click();

Reviewer 2: ```suggestion
Form().find(Button('Submit')).click();
```

> Form.find(Button(‘Submit')).click();

Reviewer: > ‘Submit'

typo

```

But if the Form component changes and now the button that submits is no longer `Submit` but `Send` we’d have to update all our tests to reflect that change. Instead we can encapsulate that detail in the `submit` action of `Form` and delegate the click action to the responsible party. Instead of reaching into internal component elements, who ever is using Form to test the form simply as:

```js

> Form.submit();

Reviewer 2: ```suggestion
Form().submit();
```

  </TabItem>
</Tabs>

The [TextField](https://github.com/thefrontside/bigtest/blob/v0/packages/interactor/src/definitions/text-field.ts) interactor from BigTest can do a lot more than this, but the example we just wrote is a good place to start when it comes to understanding how to use `createInteractor`.

Check out the source code of [createInteractor()](https://github.com/thefrontside/bigtest/blob/v0/packages/interactor/src/create-interactor.ts) for more details.

## Writing a more complex interactor

> One of the greatest benefits of Interactors is that you can turn complex testing scenarios into readable assertions. Let’s illustrate how that looks like with a commonly cumbersome UI piece: a table. We want to be able to easily assert the contents of our tables, and that means that we should be able to know the value in a cell given its column and row. Once we’re done creating a TableCell Interactor, we’ll be able to make that happen and have it look like this:

Reviewer 2: ```suggestion
One of the greatest benefits of Interactors is that you can turn complex testing scenarios into readable assertions. Let’s illustrate how that looks like with an historically cumbersome UI piece: a table. We want to be able to easily assert the contents of our tables, and that means that we should be able to know the value in a cell given its column and row. Once we’re done creating a TableCell Interactor, we’ll be able to make that happen and have it look like this:
```

```js
TableCell({ columnTitle: 'Name', rowNumber: 2 }).has({ value: 'Marge Simpson' });
```

> First, consider a table markup defined as follows:

Reviewer 2: ```suggestion
First, consider some table markup defined as follows:
```

If the built-in DOM Interactors work for your use case, they are probably the best choice.
They are maintenance-free and support the most common user actions.

When the built-in Interactors are not enough, we encourage you to write your own. They will help prevent duplicated logic in your tests, and if your interface changes, you only need to make changes to the Interactor instead of throughout the code.

For example, let's say that you want to replace a custom datepicker with a popular third-party library instead. Although you may have many tests for flows with the date picker, only your Interactor needs to change.

### I have an interaction that is really difficult to test. What should I do?

A good test suite helps you uncover hidden problems.

> Often, difficult UI tests are your early warning system for areas of your app that may have accessibility issues.

Reviewer: You can combine these two lines into one. Just a formatting thing. I think Jen likes to create new lines often when she works.

The first step is to see if you can go through the interaction yourself in the browser by using only [keyboard navigation](https://webaim.org/techniques/keyboard/).
If you cannot get to the end successfully, then you just found a bug in your app.
Although many people navigate an interface by sight and clicking,
others may use assistive technology such as screen readers and keyboard support is critical.
For example, a click can mistakenly be attached to a `div` instead of a button.

> Those types of errors can make your app unusable to some people and also difficult to test.

Reviewer: Same here.
